File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/data/JsonDocument.java
package net.ravendb.abstractions.data;

import java.util.Date;

import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.abstractions.json.linq.RavenJValue;
import net.ravendb.abstractions.util.NetDateFormat;


/**
 * A document representation:
 * - Data / Projection
 * - Etag
 * - Metadata
 */
public class JsonDocument implements IJsonDocumentMetadata {
  private RavenJObject dataAsJson;
  private RavenJObject metadata;
  private String key;
  private Boolean nonAuthoritativeInformation;
  private Etag etag;
  private Date lastModified;
  private Float tempIndexScore;

  public JsonDocument(RavenJObject dataAsJson, RavenJObject metadata, String key, Boolean nonAuthoritativeInformation, Etag etag, Date lastModified) {
    super();
    this.dataAsJson = dataAsJson;
    this.metadata = metadata;
    this.key = key;
    this.nonAuthoritativeInformation = nonAuthoritativeInformation;
    this.etag = etag;
    this.lastModified = lastModified;
  }

  /**
   * @return the dataAsJson
   */
  public RavenJObject getDataAsJson() {
    return dataAsJson != null ? dataAsJson : new RavenJObject();
  }

  /**
   * @return the etag
   */
  @Override
  public Etag getEtag() {
    return etag;
  }



  /**
   * @return the key
   */
  @Override
  public String getKey() {
    return key;
  }

  /**
   * @return the lastModified
   */
  @Override
  public Date getLastModified() {
    return lastModified;
  }

  /**
   * @return the metadata
   */
  @Override
  public RavenJObject getMetadata() {
    if (metadata == null) {
      metadata = new RavenJObject(String.CASE_INSENSITIVE_ORDER);
    }
    return metadata;
  }

  public Float getTempIndexScore() {
    return tempIndexScore;
  }

  /**
   * @return the nonAuthoritativeInformation
   */
  @Override
  public Boolean getNonAuthoritativeInformation() {
    return nonAuthoritativeInformation;
  }

  /**
   * @param dataAsJson the dataAsJson to set
   */
  public void setDataAsJson(RavenJObject dataAsJson) {
    this.dataAsJson = dataAsJson;
  }

  /**
   * @param etag the etag to set
   */
  @Override
  public void setEtag(Etag etag) {
    this.etag = etag;
  }

  /**
   * @param key the key to set
   */
  @Override
  public void setKey(String key) {
    this.key = key;
  }

  /**
   * @param lastModified the lastModified to set
   */
  @Override
  public void setLastModified(Date lastModified) {
    this.lastModified = lastModified;
  }

  /**
   * @param metadata the metadata to set
   */
  @Override
  public void setMetadata(RavenJObject metadata) {
    this.metadata = metadata;
  }

  /**
   * @param nonAuthoritativeInformation the nonAuthoritativeInformation to set
   */
  @Override
  public void setNonAuthoritativeInformation(Boolean nonAuthoritativeInformation) {
    this.nonAuthoritativeInformation = nonAuthoritativeInformation;
  }

  public void setTempIndexScore(Float tempIndexScore) {
    this.tempIndexScore = tempIndexScore;
  }

  public RavenJObject toJson() {
    dataAsJson.ensureCannotBeChangeAndEnableShapshotting();
    metadata.ensureCannotBeChangeAndEnableShapshotting();

    RavenJObject doc = dataAsJson.createSnapshot();
    RavenJObject metadata = this.metadata.createSnapshot();

    if (lastModified != null) {
      NetDateFormat fdf = new NetDateFormat();
      metadata.add(Constants.LAST_MODIFIED, new RavenJValue(this.lastModified));
      metadata.add(Constants.RAVEN_LAST_MODIFIED, new RavenJValue(fdf.format(this.lastModified)));
    }
    if (etag != null) {
      metadata.add("@etag", new RavenJValue(etag.toString()));
    }
    if (nonAuthoritativeInformation) {
      metadata.add("Non-Authoritative-Information", new RavenJValue(nonAuthoritativeInformation));
    }
    doc.add("@metadata", metadata);

    return doc;
  }

  /* (non-Javadoc)
   * @see java.lang.Object#toString()
   */
  @Override
  public String toString() {
    return "JsonDocument [dataAsJson=" + dataAsJson + ", metadata=" + metadata + ", key=" + key + ", etag=" + etag + ", lastModified=" + lastModified + "]";
  }

}

File: src/Raven.Client/Data/JsonDocument.cs
//-----------------------------------------------------------------------
// <copyright file="JsonDocument.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------

using System;
using System.Runtime.CompilerServices;
using Raven.Abstractions.Data;
using Raven.Abstractions.Extensions;
using Raven.Json.Linq;

namespace Raven.Client.Data
{
    /// <summary>
    /// A document representation:
    /// * Data / Projection
    /// * Etag
    /// * Metadata
    /// </summary>
    public class JsonDocument : IJsonDocumentMetadata
    {
        /// <summary>
        /// Create a new instance of JsonDocument
        /// </summary>
        public JsonDocument()
        {
        }

        private RavenJObject dataAsJson;
        private RavenJObject metadata;

        /// <summary>
        /// Document data or projection as json.
        /// </summary>
        public RavenJObject DataAsJson
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get { return dataAsJson ?? (dataAsJson = new RavenJObject()); }
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            set { dataAsJson = value; }
        }

        /// <summary>
        /// Metadata for the document
        /// </summary>		
        public RavenJObject Metadata
        {
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            get { return metadata ?? (metadata = new RavenJObject(StringComparer.OrdinalIgnoreCase)); }
            [MethodImpl(MethodImplOptions.AggressiveInlining)]
            set { metadata = value; }
        }

        /// <summary>
        /// Key for the document
        /// </summary>
        public string Key { get; set; }

        /// <summary>
        /// Current document etag.
        /// </summary>
        public long? Etag { get; set; }

        /// <summary>
        /// Last modified date for the document
        /// </summary>
        public DateTime? LastModified { get; set; }

        /// <summary>
        /// The ranking of this result in the current query
        /// </summary>
        public float? TempIndexScore { get; set; }

        /// <summary>
        /// How much space this document takes on disk
        /// Only relevant during indexing phases, and not available on the client
        /// </summary>
        public int SerializedSizeOnDisk;

        /// <summary>
        /// Whatever this document can be skipped from delete
        /// Only relevant during indexing phases, and not available on the client
        /// </summary>
        public bool SkipDeleteFromIndex;

        /// <summary>
        /// Translate the json document to a <see cref = "RavenJObject" />
        /// </summary>
        public RavenJObject ToJson(bool checkForId = false)
        {
            DataAsJson.EnsureCannotBeChangeAndEnableSnapshotting();
            Metadata.EnsureCannotBeChangeAndEnableSnapshotting();

            var doc = (RavenJObject)DataAsJson.CreateSnapshot();
            var metadata = (RavenJObject)Metadata.CreateSnapshot();

            if (LastModified != null)
            {
                metadata[Constants.Headers.LastModified] = LastModified.Value;
                metadata[Constants.Headers.RavenLastModified] = LastModified.Value.GetDefaultRavenFormat();
            }
            if (Etag != null)
                metadata["@etag"] = Etag.ToString();
            if (checkForId && metadata.ContainsKey("@id") == false)
                metadata["@id"] = Key;
            doc["@metadata"] = metadata;

            return doc;
        }

        public override string ToString()
        {
            return Key;
        }

        public static void EnsureIdInMetadata(IJsonDocumentMetadata doc)
        {
            if (doc == null || doc.Metadata == null)
                return;

            if (doc.Metadata.IsSnapshot)
            {
                doc.Metadata = (RavenJObject)doc.Metadata.CreateSnapshot();
            }

            doc.Metadata["@id"] = doc.Key;
        }
    }
}
